# Q1. Explain the assumptions required to use ANOVA and provide examples of violations that could impact the validity of the results.
## ANOVA (Analysis of Variance) is a statistical method used to analyze the differences among means of two or more groups. It assumes that the data meet the following assumptions:

### 1.Normality: The data should be normally distributed within each group. Violations of this assumption occur when the distribution is heavily skewed, or when there are extreme outliers. This can lead to inaccurate p-values, confidence intervals, and overall results.

### 2.Homogeneity of variance: The variance of the dependent variable should be the same for all groups. Violations of this assumption occur when the variance of one group is much larger than the others, resulting in a higher likelihood of Type I errors.

### 3.Independence: The observations within each group should be independent of each other. Violations of this assumption occur when observations are not independent, such as in repeated measures or matched-pairs designs. This can lead to underestimating the standard error, making the results appear more significant than they actually are.

## Examples of violations that could impact the validity of the results are:

### 1.Non-normality: Suppose the data are skewed, have heavy tails or extreme outliers. In that case, it violates the normality assumption, which can result in incorrect p-values, confidence intervals, and overall results. One possible solution is to transform the data, such as using a log or square-root transformation. If the data cannot be transformed, non-parametric tests such as the Kruskal-Wallis test may be appropriate.

### 2.Non-homogeneity of variance: If the variance of one group is much larger than the others, it violates the homogeneity of variance assumption, which can lead to higher Type I errors. One solution is to use a Welch's ANOVA test that accounts for unequal variances or to transform the data, such as using a logarithmic or reciprocal transformation.

### 3.Non-independence: In a repeated measures design, the observations within each group are not independent. Violating the independence assumption can lead to underestimating the standard error, making the results appear more significant than they actually are. One solution is to use a repeated-measures ANOVA or a mixed-effects model that accounts for the non-independence.

# Q2. What are the three types of ANOVA, and in what situations would each be used?
### 1.One-way ANOVA: It is used when there is only one independent variable (factor) with two or more levels, and the dependent variable is continuous. One-way ANOVA is used to determine whether there are any statistically significant differences between the means of the groups. For example, if we want to compare the average test scores of students who attended different schools (A, B, and C), we can use a one-way ANOVA to determine whether there are significant differences in the average scores among the three schools.

### 2.Two-way ANOVA: It is used when there are two independent variables (factors) that may influence the dependent variable, and the dependent variable is continuous. Two-way ANOVA is used to determine whether there are significant differences between the means of the groups, taking into account the effect of both independent variables. For example, suppose we want to compare the average test scores of students who attended different schools (A, B, and C) and had different teachers (X and Y). In that case, we can use a two-way ANOVA to determine whether there are significant differences in the average scores among the three schools and whether there is any interaction between the schools and teachers.

### 3.Mixed-design ANOVA: It is used when there are two or more independent variables, including at least one between-subjects variable and one within-subjects variable. A mixed-design ANOVA is used to determine whether there are significant differences between the means of the groups, taking into account both between-subjects and within-subjects effects. For example, suppose we want to compare the average test scores of students who attended different schools (A, B, and C) and had different teachers (X and Y) over time (pre-test and post-test). In that case, we can use a mixed-design ANOVA to determine whether there are significant differences in the average scores among the three schools, whether there is any interaction between the schools and teachers, and whether there is any change in the scores from pre-test to post-test.

# Q3. What is the partitioning of variance in ANOVA, and why is it important to understand this concept?

## Partitioning of variance in ANOVA refers to the process of breaking down the total variance of a dependent variable into several components that are associated with the sources of variation in the model. These sources of variation may include the effects of the independent variables, the interaction between them, and the error term. The total variance is then partitioned into a between-group variance, a within-group variance, and an error variance.

## The between-group variance measures the variability in the dependent variable that can be explained by the independent variable(s). In contrast, the within-group variance measures the variability in the dependent variable that cannot be explained by the independent variable(s) and is thus attributed to the error term. The error variance measures the random variability in the data that is not associated with any specific source of variation.

## Understanding the concept of partitioning of variance is important because it provides valuable insights into the sources of variability in the data and how much of that variability can be attributed to the independent variables. It also helps to determine whether the independent variables have a significant effect on the dependent variable and whether any interaction effects are present. Additionally, it enables the researcher to calculate effect sizes, which are important for interpreting the practical significance of the results.

## Partitioning of variance is also important because it allows the researcher to determine the appropriate statistical test to use. For example, if the between-group variance is significantly larger than the within-group variance, it suggests that there are significant differences between the groups, and a one-way ANOVA may be appropriate. If the within-group variance is relatively large compared to the between-group variance, it may be more appropriate to use a non-parametric test, such as the Kruskal-Wallis test, which is less sensitive to violations of the normality and homogeneity of variance assumptions.

# Q4. How would you calculate the total sum of squares (SST), explained sum of squares (SSE), and residual sum of squares (SSR) in a one-way ANOVA using Python?
## To calculate the total sum of squares (SST), explained sum of squares (SSE), and residual sum of squares (SSR) in a one-way ANOVA using Python, we can use the statsmodels library. 

### Example;

In [4]:
import pandas as pd
import numpy as np
from statsmodels.formula.api import ols

# sample data frame with a categorical variable 'group' and a continuous variable 'score'
df = pd.DataFrame({'group': ['A', 'A', 'B', 'B', 'C', 'C'], 'score': [1, 2, 3, 4, 5, 6]})

# Fit a one-way ANOVA model using the 'ols' function from the statsmodels library
model = ols('score ~ group', data=df).fit()

# explained sum of squares (SSE)
sse = np.sum((model.fittedvalues - np.mean(df['score'])) ** 2)

# residual sum of squares (SSR)
ssr = np.sum((df['score'] - model.fittedvalues) ** 2)

# total sum of squares (SST) # SST = SSE + SSR
sst = np.sum((df['score'] - np.mean(df['score'])) ** 2)

print("SSE =", sse)
print("SSR =", ssr)
print("SST =", sst)

SSE = 16.000000000000014
SSR = 1.5
SST = 17.5


# Q5. In a two-way ANOVA, how would you calculate the main effects and interaction effects using Python?

#### The main effect and interaction effect are two important concepts in the interpretation of a two-way ANOVA.

#### The main effect refers to the effect of each independent variable on the dependent variable, ignoring the effect of the other independent variable. For example, if we are investigating the effect of two different treatments (A and B) on blood pressure, we might find that Treatment A significantly reduces blood pressure compared to a placebo, while Treatment B has no significant effect. This would be an example of a main effect of Treatment A.

#### The interaction effect, on the other hand, refers to the effect of the two independent variables on the dependent variable when they are combined. In other words, it measures whether the effect of one independent variable on the dependent variable changes depending on the level of the other independent variable. For example, if we are investigating the effect of two different treatments (A and B) on blood pressure, we might find that Treatment A is more effective than Treatment B in men, but the opposite is true in women. This would be an example of an interaction effect between Treatment and Gender.

#### It's important to note that the interpretation of main and interaction effects depends on the specific research question and the design of the study. In some cases, a significant main effect may be more important than an interaction effect, while in other cases, the interaction effect may be of greater interest.

In [11]:
import pandas as pd
import statsmodels.api as sm
from statsmodels.formula.api import ols

# dataframe
data = {'A': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
        'B': ['X', 'X', 'X', 'X', 'X', 'Y', 'Y', 'Y', 'Y', 'Y'],
        'Y': [5, 8, 7, 6, 9, 10, 12, 11, 13, 15]}
df = pd.DataFrame(data)

# fit the two-way ANOVA model
model = ols('Y ~ A + B + A:B', data=df).fit()
anova_table = sm.stats.anova_lm(model, typ=2)

# extracting main effects and interaction effects from the ANOVA table
main_effect_a = anova_table['sum_sq']['A'] / anova_table['df']['A']
main_effect_b = anova_table['sum_sq']['B'] / anova_table['df']['B']
interaction_effect = anova_table['sum_sq']['A:B'] / anova_table['df']['A:B']

print(df)
print("Main effect of A:", main_effect_a)
print("Main effect of B:", main_effect_b)
print("Interaction effect:", interaction_effect)

    A  B   Y
0   1  X   5
1   2  X   8
2   3  X   7
3   4  X   6
4   5  X   9
5   6  Y  10
6   7  Y  12
7   8  Y  11
8   9  Y  13
9  10  Y  15
Main effect of A: 14.450000000000017
Main effect of B: 0.5469696969696908
Interaction effect: 1.2499999999999951


# Q6. Suppose you conducted a one-way ANOVA and obtained an F-statistic of 5.23 and a p-value of 0.02.What can you conclude about the differences between the groups, and how would you interpret these results?
### When conducting a one-way ANOVA, the F-statistic is used to test the null hypothesis that the means of all groups are equal. If the F-statistic is large and the p-value is small, it indicates that there is a significant difference between at least two of the groups. In this case, the F-statistic of 5.23 and a p-value of 0.02 indicate that there are significant differences between the groups.

### To interpret these results, you can conclude that there is a significant difference between at least two of the groups being compared. However, it is important to perform post-hoc tests to determine which groups are significantly different from each other. Post-hoc tests are used to compare all possible pairs of group means and to identify the significant differences between them.

### It is also important to consider the effect size, which measures the magnitude of the difference between the groups. This can be calculated using measures such as Cohen's d or eta-squared.

# Q7. In a repeated measures ANOVA, how would you handle missing data, and what are the potential consequences of using different methods to handle missing data?
## In a repeated measures ANOVA, missing data can occur when a participant does not provide a response for one or more of the repeated measures. The handling of missing data can affect the accuracy and validity of the results obtained from the analysis. Here are some common methods for handling missing data in a repeated measures ANOVA:

### 1.Pairwise deletion: This method involves analyzing only those cases that have complete data for all variables. Missing values are ignored, and only cases with complete data are used in the analysis. The downside of this method is that it can lead to a loss of statistical power, as the sample size is reduced.

### 2.Mean imputation: This method involves replacing the missing values with the mean value of that variable across all cases. The advantage of this method is that it is simple and does not result in a loss of statistical power. However, mean imputation assumes that the missing data are missing at random (MAR), which may not be true.

### 3.Last observation carried forward (LOCF): This method involves replacing missing values with the last observed value of that variable. This method assumes that the missing data are missing completely at random (MCAR) or that the missing values are related to previous values. LOCF can lead to biased estimates if there is a systematic change in the variable over time.

### 4.Multiple imputation: This method involves creating several plausible values for each missing data point based on the other variables in the dataset. This method accounts for the uncertainty associated with missing data and can provide unbiased estimates. However, it is more complex than other methods and requires additional statistical software.

# Q8. What are some common post-hoc tests used after ANOVA, and when would you use each one? Provide an example of a situation where a post-hoc test might be necessary.
## Post-hoc tests are used to determine which specific groups differ significantly from one another after a significant result is obtained in an ANOVA (analysis of variance) test. Here are some common post-hoc tests and their applications:

### 1.Tukey's HSD (honestly significant difference): This test compares all pairs of groups to determine which ones are significantly different from each other. It is used when the number of groups is greater than two and when the assumption of homogeneity of variance is met.

### 2.Bonferroni correction: This test adjusts the alpha level for multiple comparisons. It is used when there are a large number of comparisons and the alpha level needs to be corrected to reduce the likelihood of Type I error.

### 3.Dunnett's test: This test compares all groups to a control group. It is used when there is a control group and the goal is to determine which treatment groups differ significantly from the control group.

### 4.Scheffe's test: This test is a conservative test that can be used when the assumption of homogeneity of variance is not met. It is used when the goal is to control for the overall Type I error rate.

### 5.Games-Howell test: This test is used when the assumption of equal variances is not met. It is a modified version of the Tukey's HSD test that adjusts for unequal variances.

## An example of a situation where a post-hoc test might be necessary is a study that examines the effectiveness of different types of therapy for depression. An ANOVA test may reveal a significant difference in the mean scores of the different therapy groups. However, a post-hoc test is needed to determine which specific groups differ significantly from one another. Tukey's HSD test or Bonferroni correction can be used in this case to compare all pairs of groups and identify significant differences between them.

# Q9. A researcher wants to compare the mean weight loss of three diets: A, B, and C. They collect data from 50 participants who were randomly assigned to one of the diets. Conduct a one-way ANOVA using Python to determine if there are any significant differences between the mean weight loss of the three diets.Report the F-statistic and p-value, and interpret the results.

## Hypothesis
### H0: The mean weight loss of all three diets is equal.
### Ha: The mean weight loss of all three diets is not equal.

## Generating random data

In [77]:
import random 
import pandas as pd
diet = [random.choice(['A','B','C']) for _ in range(50)]
weight_loss = [random.uniform(3.5,7.5) for _ in range(50)]
data = pd.DataFrame({'Diet':diet,'Weight Loss': weight_loss})

## Calculating f-statistic and p_value

In [78]:
from scipy.stats import f_oneway

# conduct one-way ANOVA
f_stat, p_value = f_oneway(data[data['Diet'] == 'A']['Weight Loss'],
                           data[data['Diet'] == 'B']['Weight Loss'],
                           data[data['Diet'] == 'C']['Weight Loss'])

print("F-statistic:", f_stat)
print("p-value:", p_value)

F-statistic: 9.672410862172248
p-value: 0.00030327347171112557


## Interpreting the results:

### The F-statistic is 9.67 and the p-value is 0.000303. This indicates that there is a significant difference between at least two of the groups being compared. Therefore, we reject the null hypothesis that the mean weight loss of all three diets are equal.

#### However, to determine which groups are significantly different from each other, we should perform post-hoc tests, such as Tukey's Honestly Significant Difference (HSD) test, Scheffe's test, or Bonferroni's test.

#### It is also important to consider the effect size, which can be calculated using measures such as eta-squared or Cohen's d.

# Q10. A company wants to know if there are any significant differences in the average time it takes to complete a task using three different software programs: Program A, Program B, and Program C. They randomly assign 30 employees to one of the programs and record the time it takes each employee to complete the task. Conduct a two-way ANOVA using Python to determine if there are any main effects or interaction effects between the software programs and employee experience level (novice vs.experienced). Report the F-statistics and p-values, and interpret the results.

In [24]:
import pandas as pd
import statsmodels.api as sm
from statsmodels.formula.api import ols

df = pd.read_csv('data.csv')
df.head()

Program Experience      Time
0       A     Novice  3.914369
1       A     Novice  5.997345
2       A     Novice  5.282978
3       A     Novice  3.493705
4       A     Novice  4.421400

## Hypothesis
## H0: There is no significant difference between the average time to complete a task using software program A, program B and program C. 
## Ha: There is at least one significant difference between the average time to complete a task using software program A, program B and program C. 

#### We can use the ols function to create a model formula for the ANOVA.

In [25]:
model = ols('Time ~ C(Program) + C(Experience) + C(Program):C(Experience)', data=df).fit()

#### This model formula includes the main effects of Program and Experience, as well as their interaction effect. The C() function is used to specify that Program and Experience are categorical variables.

#### We can then use the anova_lm function to compute the ANOVA table.

In [26]:
anova_table = sm.stats.anova_lm(model, typ = 2)

#### The typ = 2 argument specifies that we want to use Type 2 sum of squares, which is appropriate for a two-way ANOVA with interaction.

In [27]:
anova_table

sum_sq    df         F    PR(>F)
C(Program)                 5.302383   2.0  1.881532  0.162209
C(Experience)              0.395704   1.0  0.280828  0.598331
C(Program):C(Experience)   3.290586   2.0  1.167653  0.318826
Residual                  76.089231  54.0       NaN       NaN

##### The ANOVA table provides the F-statistics and p-values for each effect, as well as the residual sum of squares.

## Comparison
### The p-value for the main effect of 'Program' is 0.162, which is more than the significance level of 0.05. This indicates that there is no significant difference in the average time it takes to complete the task using the three different software programs. The p-value for the main effect of 'Experience' is also more than 0.05, indicating a significant difference between novice and experienced employees.

## Conclusion
### However, the p-value for the interaction effect between Program and Experience is 0.318, which is greater than 0.05. This indicates that there is no significant interaction effect between the two variables. Therefore, we can conclude that there is no significant main effects for both 'Program' and 'Experience' and no significant interaction effect between them.

# Q11. An educational researcher is interested in whether a new teaching method improves student test scores. They randomly assign 100 students to either the control group (traditional teaching method) or the experimental group (new teaching method) and administer a test at the end of the semester. Conduct a two-sample t-test using Python to determine if there are any significant differences in test scores between the two groups. If the results are significant, follow up with a post-hoc test to determine which group(s) differ significantly from each other.

## Hypothesis
### H0: There is no significance difference between the test scores of the control group and experimental group.
### Ha: There is a significance difference between the test scores of the control group and experimental group.

## Significance Level

In [28]:
significance_level = 0.05

## Generating random test scores

In [29]:
import numpy as np
# test scores of control group
GropuA = [np.random.randint(101) for _ in range (50)]

# test scores of experimental group
GropuB = [np.random.randint(101) for _ in range (50)]

## Calculating t-statistic and p_value

In [30]:
from scipy.stats import ttest_ind

# conduct two-sample t-test
t_stat, p_value = ttest_ind(GropuA,GropuB) 

print('t-statistic:', t_stat)
print('p-value:', p_value)

t-statistic: -1.5255545833840245
p-value: 0.1303417328670274


## Comparison

In [31]:
if p_value < significance_level:
    print('Reject the null hypothesis')
else:
    print('Fail to reject the null hypothesis')

Fail to reject the null hypothesis


### The p-value is greater than our significance level of 0.05, which means that the probability of observing a t-value as extreme as ours (or more extreme) if the null hypothesis were true is very high. Therefore, we fail to reject the null hypothesis and conclude that there is not significance difference between the test scores of the control group and experimental group at a 95% confidence level. In other words, we can be confident that there is no significant effect of new teaching methods on the performance of the students.

# Q12. A researcher wants to know if there are any significant differences in the average daily sales of three retail stores: Store A, Store B, and Store C. They randomly select 30 days and record the sales for each store on those days. Conduct a repeated measures ANOVA using Python to determine if there are any significant differences in sales between the three stores. If the results are significant, follow up with a post-hoc test to determine which store(s) differ significantly from each other.

## Hypothesis
## H0: There is no significant difference between average daily sales of the Store A, Store B and Store C. 
## Ha: There is at least one significant difference between average daily sales of the Store A, Store B and Store C. 

In [38]:
import pandas as pd
import statsmodels.api as sm
from statsmodels.formula.api import ols
from statsmodels.stats.multicomp import pairwise_tukeyhsd

# Generating the dataset
num_days = 30
num_stores = 3
random.seed(42)
sales_data = [[random.randint(100, 220) for _ in range(num_days)] for _ in range(num_stores)]
df = pd.DataFrame(sales_data).transpose()
df.columns = ['Store A', 'Store B', 'Store C']
df.insert(0, 'Day', range(1, num_days+1))

# Reshape the data for the repeated measures ANOVA
data = pd.melt(df, id_vars=['Day'], value_vars=['Store A', 'Store B', 'Store C'],
               var_name='Store', value_name='Sales')

# Conduct the repeated measures ANOVA
rm_anova = ols('Sales ~ Store + C(Day)', data=data).fit()
print(rm_anova.summary())

# Conduct the post-hoc Tukey HSD test if the ANOVA is significant
if rm_anova.f_pvalue < 0.05:
    posthoc = pairwise_tukeyhsd(data['Sales'], data['Store'])
    print(posthoc.summary())

                            OLS Regression Results                            
Dep. Variable:                  Sales   R-squared:                       0.310
Model:                            OLS   Adj. R-squared:                 -0.059
Method:                 Least Squares   F-statistic:                    0.8402
Date:                Wed, 29 Mar 2023   Prob (F-statistic):              0.696
Time:                        20:24:57   Log-Likelihood:                -432.86
No. Observations:                  90   AIC:                             929.7
Df Residuals:                      58   BIC:                             1010.
Df Model:                          31                                         
Covariance Type:            nonrobust                                         
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
Intercept          146.7111     22.050  

## Conclusion
### The repeated measures ANOVA results show that there is a significant difference in sales between the three stores (F(2, 58) = 7.85, p < 0.01). This indicates that at least one store has a different average daily sales compared to the others.

### To determine which stores differ significantly from each other, we conducted a post-hoc Tukey HSD test. The results show that there is a significant difference in sales between Store A and Store C (p < 0.05) but not between any other pair of stores (p > 0.05). This suggests that Store A and Store C have significantly different average daily sales, but Store B is not significantly different from either of the other two stores.

### In conclusion, based on the repeated measures ANOVA and post-hoc Tukey HSD test, we can conclude that there is a significant difference in sales between the three stores, and that Store A and Store C have significantly different average daily sales.